third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1184 @@
+//      a local variable conflicts with that variable.  It uses `SymbolPool::shadow` to create
+//      shadow symbols which print out the same but don't conflict.
+//
+//   - `get_shadow_symbol` checks whether a `Symbol` conflicts with lambda free variables, and if so,
+//     returns the shadow symbol to use instead..
+//
+//   - `parameter_list_to_pattern` is a helper to convert a list of `Parameter` into a `Pattern`,
+//      suitable for use in the body.
+//
+//   - `construct_inlined_call_expression` is a helper to build the expression corresponding to
+//      { let params=actuals; body } used for both lambda inlining and inline function inlining.
+//
+//   - rewrite_pattern_vector is a helper for `rewrite_pattern`
+//
+
+// TODO:
+// - add a InlinedCall() node that represents an inlined call, so that an inlined Return() has a place to go. [later]
+//   - OR: if an inlined fn has a return, then error out.
+// - add a simplifier that simplifies certain code constructs, e.g.
+//   - InlinedCall() with no nested Return() can be flattened.
+//   - others?
+// - do we need to insert FreezeRef in actual args if formal param is &T but arg is &mod T ?
+// - do we need to do anything about abilities?
+// - if lambda parameters also may be referred to by temporaries, then `rewrite_invoke` might need to call yet
+//   another `ExpRewriterFunctions` implementation to do that.
+
+use codespan_reporting::diagnostic::Severity;
+use itertools::chain;
+use move_model::{
+    ast::{Exp, ExpData, Operation, Pattern, TempIndex},
+    exp_rewriter::ExpRewriterFunctions,
+    model::{FunId, GlobalEnv, Loc, NodeId, Parameter, QualifiedId, TypeParameter},
+    symbol::Symbol,
+    ty::{ReferenceKind, Type},
+};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fmt::Debug,
+    iter::{zip, IntoIterator, Iterator},
+    ops::Deref,
+    vec::Vec,
+};
+
+type QualifiedFunId = QualifiedId<FunId>;
+type CallSiteLocations = BTreeMap<(QualifiedFunId, QualifiedFunId), BTreeSet<NodeId>>;
+
+// ======================================================================================
+// Entry
+
+// Run inlining on current program's AST.  For each function which is target of the compilation,
+// visit that function body and inline any calls to functions marked as "inline".
+// debug flag is temporary until we sort out a way to attach debug to GlobalEnv.
+pub fn run_inlining(env: &mut GlobalEnv, debug: bool) {
+    // Start with targets.
+    let mut todo = BTreeSet::new();
+    // While we're iterating, error on any target inline functions lacking a body to inline.
+    let mut has_error: bool = false;
+    for module in env.get_modules() {
+        if module.is_target() {
+            for func in module.get_functions() {
+                let id = func.get_qualified_id();
+                if func.is_inline() {
+                    let func_def = func.get_def();
+                    if func_def.deref().is_none() {
+                        has_error = true;
```

Comment:
`func_dev` is a `Ref<Option<Exp>>`.  At some point the `.deref()` seemed necessary, and nothing was complaining about it, but now I see it builds without it in this spot.

Reply:
`Ref<T>` automatically casts to `&T` in some situations, for example with receiver notation. It does this via the `AsRef<T>` trait. Otherwise, it is actually idiomatic to do `&*e` (at least what I have seen and do).

Reply:
Done.